enhancements/kube-apiserver/stability-api-removal-notifications.md

```diff
@@ -0,0 +1,162 @@
+---
+title: api-removal-notifications
+authors:
+  - "@sanchezl"
+reviewers:
+  - "@deads2k"
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-02-01
+last-updated: 2021-02-01
+status: provisional
+---
+
+# API Removal Notifications
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Notify customers that an API that will be removed in the next release is in use.
+
+## Motivation
+
+Kubernetes is being more aggressive with the removal of APIs that have been deprecated or pre-release APIs that have
+reached a dead-end and will never reach GA. Additionally, the ability to re-enable a previously removed API has been
+removed.
+
+Customers should be notified that an API that will be removed in the next release is in use so that they can decide if
+there is a need to take action in order to ensure success when upgrading to the next release.
+
+### Goals
+
+1. Notify customers that an API that will be removed in the next release is in use.
+2. Notify EUS customers that an API that will be removed in the next EUS release is in use.
+3. Document how to identify the workload that is using the API to be removed.
+
+### Non-Goals
+
+1. Automatically identify the workload using the API to be removed.
+2. Prevent the customer from attempting to upgrade when an API that will be removed in the next release is in use.
```

Comment:
I think this should become a goal, if an object just vanishes during upgrade then we should inhibit that upgrade. So Info level alerts on anything removed in N through N+3, Upgradeable=False on removal in N+1.

Reply:
> I think this should become a goal, if an object just vanishes during upgrade then we should inhibit that upgrade. So Info level alerts on anything removed in N through N+3, Upgradeable=False on removal in N+1.

I'm willing to debate that on a larger arch call, but without the ability to selectively override conditions for an upgrade, such a condition is likely to become a regularly performed and indiscriminately used --force.

For the purposes of this enhancement, an non-goal is a very reasonable starting point to get the majority of development done.